python/cudf_polars/cudf_polars/dsl/nodebase.py

```diff
@@ -0,0 +1,120 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Base class for IR nodes, and utilities."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, ClassVar
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from typing_extensions import Self
+
+
+class Node:
+    """
+    An abstract node type.
+
+    Nodes are immutable!
+
+    This contains a (potentially empty) tuple of child nodes,
+    along with non-child data. For uniform reconstruction and
+    implementation of hashing and equality schemes, child classes need
+    to provide a certain amount of metadata when they are defined.
+    Specifically, the ``_non_child`` attribute must list, in-order,
+    the names of the slots that are passed to the constructor. The
+    constructor must take arguments in the order ``(*_non_child,
+    *children).``
+    """
+
+    __slots__ = ("_hash_value", "_repr_value", "children")
+    _hash_value: int
+    _repr_value: str
+    children: tuple[Node, ...]
+    _non_child: ClassVar[tuple[str, ...]] = ()
+
+    def _ctor_arguments(self, children: Sequence[Node]) -> Sequence:
+        return (*(getattr(self, attr) for attr in self._non_child), *children)
+
+    def reconstruct(
+        self, children: Sequence[Node]
+    ) -> Self:  # pragma: no cover; not yet used
+        """
+        Rebuild this node with new children.
+
+        Parameters
+        ----------
+        children
+            New children
+
+        Returns
+        -------
+        New node with new children. Non-child data is shared with the input.
+        """
+        return type(self)(*self._ctor_arguments(children))
+
+    def get_hash(self) -> int:
+        """Return a hash of the node."""
+        return hash((type(self), self._ctor_arguments(self.children)))
```

Comment:
Hash values are likely to be used often in rewriting and (eventually) task-graph generation. You are implementing caching within `__hash__`. I wonder if it makes more sense to do the caching right here, and to discourage subclasses from re-implementing `get_hash`.

In order to avoid the need for multiple `get_hash` definitions, we could possibly use a dispatching the same way dask does for "tokenization":

```python
    def get_hash(self) -> int:
        """Return a hash of the node."""
         try:
            return self._hash_value
        except AttributeError:
            self._hash_value = hash(
                (
                    type(self),
                    *self.children,
                    *[
                        make_hashable(getattr(self, attr)) for attr in self._non_child
                    ],
                )
            )
            return self._hash_value
```

Here, the hypothetical `make_hashable` function would be a singledispatch used to "normalize" non-child arguments. That way we can define the necessary code to normalize a specific datatype only once.

Reply:
My goal is that no-one calls get_hash directly. If you want the hash of something, say hash!

This is just to allow node-specific overrides with a generic fallback. 

It's only two things in the IR nodes that need something specific, so I'm not sure if it's worth defining the dispatch mechanism
idea just for that.

Reply:
>My goal is that no-one calls get_hash directly

Okay, this sounds reasonable to me. Should the method look "private" (e.g. `_get_hash`)?

Reply:
Python’s builtin `hash` will not be consistent across workers or between runs. It is seeded differently on process startup for security reasons. Does that matter? The builtin `hash` is typically only useful for dict/set-like applications within a process, in my experience, and overloading it can be error prone.

Reply:
I only need `hash` to be stable in a process for dictionaries. I think if Rick needs something different then we should figure out how to provide that.

Reply:
For example, if we need a stable hash, I recommend we would implement (on the abstract base class):

```
def stable_hash(self) -> str:
    if self._stable_hash is not None:
        return self._stable_hash
    self._stable_hash = hashlib.md5sum(self.get_hash()).hexdigest()
    return self._stable_hash
```

Or something like that.

Note that some objects contain "heavy" data (arrow arrays or polars dataframes) and producing a hash of those in a deterministic manner is much more difficult.

Reply:
Thanks for this discussion @wence- and @bdice !

My understanding is that we want every `Node` object to be associated with a unique ID for two purposes:

1. In order to implement (caching) rewrites, we probably need unique references to these objects in dictionaries and sets. As @wence- pointed out, we only _need_ these objects to have a stable hash within a running process to satisfy this purpose.
2. When it comes time to "lower" a `Node` into a task graph, we want the output tasks for each node to use keys that are deterministically mapped to that object. In most cases it is probably "okay" if these keys are only deterministic within a single (client) process. However, that definitely makes me feel uneasy.

>I'm presuming that without persisting things in dask, once a graph is created and executed, I am allowed to reuse the same keys, and dask won't match them up with an already executed key that might happen to have the same name.

No, Dask definitely assumes there is a 1:1 mapping between keys and tasks. Therefore you cannot "re-use" a key name after persisting, because the worker will just return the already-computed result if it's still available. This is rarely a problem in practice. The more-common case is that a new key is used for a task that is already executed, and so we end up repeating the work.

>Note that some objects contain "heavy" data (arrow arrays or polars dataframes) and producing a hash of those in a deterministic manner is much more difficult.

Right exactly. This is an excellent point. This is also a problem in Dask DataFrame/Array. In order to get a deterministic token for GPU data (cudf or cupy), we literally move the data to host to calculate the token. This is **terrible** in terms of performance. However, we can rationalize it because it is actually an anti-pattern to use materialized data in the task graph. Unfortunately, this is not the case for single-gpu `cudf.polars`! 

Perhaps the best solution is to focus on requirement (1) in this PR: We just want `Node` objects to be deterministically hashable within a process. In order to satisfy the needs of Dask, we can just implement [__dask_tokenize__](https://docs.dask.org/en/stable/custom-collections.html#deterministic-hashing) on these classes, and use Dask's `tokenize` to produce keys.

Reply:
@wence- - What do you think about implementing `get_hashable(self) -> Hashable` instead of `get_hash(self) -> int`?

This way, we could return `hash(self.get_hashable())` within the `__hash__` definition. And we could return `tokenize(self.get_hashable())` within `__dask_tokenize__`. For special cases with "heavy" data, we could override the default `__dask_tokenize__` definition.

Reply:
Yes, that sounds fine.

Reply:
Thanks for the discussion, I agree with the proposed solution!